### PR TITLE
support UEFI boot with qemu/kvm

### PR DESCRIPTION
**Pull request by gqlo**
### What does this PR do?
This PR add the support for UEFI boot with qemu//kvm. Users can use `loader` parameters  to load the OVMF file.

### New Behavior
```
- boot:
        loader: /usr/share/OVMF/OVMF_CODE.fd
```

### Tests written?
**[NOTICE] Bug fixes or features added to Salt require tests.**
Please review the [test documentation](https://docs.saltstack.com/en/latest/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite.

No

### Commits signed with GPG?

Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/modules/virt.py [hunk 1]
```diff
@@ -1168,22 +1168,55 @@ def _handle_remote_boot_params(orig_boot):
...
+    rename the file and copy to the corresponding location.
+    :param name: the name of the domain
+    :param orig_boot: The original boot parameters passed to the init or update
+    :return:  string that contains the path of variable store file under qemu
+    """
+    qemu_dir = '/var/lib/libvirt/qemu/nvram'
+    if not os.path.exists(qemu_dir):
+        os.makedirs(qemu_dir)
+    dirname = os.path.dirname(orig_boot.get('loader'))
+    src = '{}/OVMF_VARS.fd'.format(dirname)
```
Comment (cbosdo) [inline salt/modules/virt.py:1202]: in some distros like openSUSE the var templates files aren't named like this. Here is the content of the `qemu-ovmf-x86_64` package on openSUSE:

```
# rpm -ql qemu-ovmf-x86_64                                                                                                                                                                       
/usr/share/licenses/qemu-ovmf-x86_64
/usr/share/licenses/qemu-ovmf-x86_64/License-ovmf.txt
/usr/share/licenses/qemu-ovmf-x86_64/License.txt
/usr/share/qemu
/usr/share/qemu/firmware
/usr/share/qemu/firmware/50-ovmf-x86_64-secure-ms.json
/usr/share/qemu/firmware/50-ovmf-x86_64-secure-opensuse.json
/usr/share/qemu/firmware/50-ovmf-x86_64-secure-suse.json
/usr/share/qemu/firmware/50-ovmf-x86_64-secure.json
/usr/share/qemu/firmware/60-ovmf-x86_64.json
/usr/share/qemu/ovmf-x86_64-4m-code.bin
/usr/share/qemu/ovmf-x86_64-4m-vars.bin
/usr/share/qemu/ovmf-x86_64-4m.bin
/usr/share/qemu/ovmf-x86_64-code.bin
/usr/share/qemu/ovmf-x86_64-ms-4m-code.bin
/usr/share/qemu/ovmf-x86_64-ms-4m-vars.bin
/usr/share/qemu/ovmf-x86_64-ms-4m.bin
/usr/share/qemu/ovmf-x86_64-ms-code.bin
/usr/share/qemu/ovmf-x86_64-ms-vars.bin
/usr/share/qemu/ovmf-x86_64-ms.bin
/usr/share/qemu/ovmf-x86_64-opensuse-4m-code.bin
/usr/share/qemu/ovmf-x86_64-opensuse-4m-vars.bin
/usr/share/qemu/ovmf-x86_64-opensuse-4m.bin
/usr/share/qemu/ovmf-x86_64-opensuse-code.bin
/usr/share/qemu/ovmf-x86_64-opensuse-vars.bin
/usr/share/qemu/ovmf-x86_64-opensuse.bin
/usr/share/qemu/ovmf-x86_64-smm-code.bin
/usr/share/qemu/ovmf-x86_64-smm-ms-code.bin
/usr/share/qemu/ovmf-x86_64-smm-ms-vars.bin
/usr/share/qemu/ovmf-x86_64-smm-ms.bin
/usr/share/qemu/ovmf-x86_64-smm-opensuse-code.bin
/usr/share/qemu/ovmf-x86_64-smm-opensuse-vars.bin
/usr/share/qemu/ovmf-x86_64-smm-opensuse.bin
/usr/share/qemu/ovmf-x86_64-smm-suse-code.bin
/usr/share/qemu/ovmf-x86_64-smm-suse-vars.bin
/usr/share/qemu/ovmf-x86_64-smm-suse.bin
/usr/share/qemu/ovmf-x86_64-smm-vars.bin
/usr/share/qemu/ovmf-x86_64-smm.bin
/usr/share/qemu/ovmf-x86_64-suse-4m-code.bin
/usr/share/qemu/ovmf-x86_64-suse-4m-vars.bin
/usr/share/qemu/ovmf-x86_64-suse-4m.bin
/usr/share/qemu/ovmf-x86_64-suse-code.bin
/usr/share/qemu/ovmf-x86_64-suse-vars.bin
/usr/share/qemu/ovmf-x86_64-suse.bin
/usr/share/qemu/ovmf-x86_64-vars.bin
/usr/share/qemu/ovmf-x86_64.bin
```

For sure we'll need to improve the heuristic to find those files from the binary one.

File: salt/modules/virt.py [hunk 2]
```diff
@@ -1168,22 +1168,55 @@ def _handle_remote_boot_params(orig_boot):
...
     new_boot = orig_boot.copy()
+    keys = orig_boot.keys()
 
     try:
-        for key in ['kernel', 'initrd']:
-            if check_remote(orig_boot.get(key)):
-                if saltinst_dir is None:
-                    os.makedirs(CACHE_DIR)
-                    saltinst_dir = CACHE_DIR
+        if boot_param_is_valid(keys):
```
Comment (cbosdo) [inline salt/modules/virt.py:1174]:
```suggestion
        if not boot_param_is_valid(keys):
            raise SaltInvocationError("Invalid boot parameter: kernel and initrd need to be both provided")
```

And then you don't change the indentation level of the code after. If the state parameters are wrong, then we need to fail and tell it to the user.

File: salt/modules/virt.py [hunk 3]
```diff
@@ -1168,22 +1168,55 @@ def _handle_remote_boot_params(orig_boot):
...
+    check if boot parameter is valid. For example, passing initrd withtout kernel would be invalid since
+    initrd is specific to the kernel.
+    :param keys: list of keys from the dictionary that are passed to boot parameters
+    :return: a boolean value indicates the validity of boot parameters
+    """
+    return keys == {'loader'} or \
+           keys == {'kernel', 'initrd'} or \
+           keys == {'kernel', 'initrd', 'loader'}
+
+
```
Comment (cbosdo) [inline salt/modules/virt.py:1219]: Since this is just a rather simple check, I wouldn't create a function for it

File: salt/modules/virt.py [hunk 4]
```diff
@@ -1358,7 +1391,7 @@ def init(name,
     .. rubric:: Network Interfaces Definitions
 
     Network interfaces dictionaries can contain the following properties:
-
+boot
```
Comment (cbosdo) [inline salt/modules/virt.py:1394]: Wrong place for this documentation item

File: salt/modules/virt.py [hunk 5]
```diff
@@ -1899,20 +1932,17 @@ def update(name,
         need_update = True
 
     # Update the kernel boot parameters
-    boot_tags = ['kernel', 'initrd', 'cmdline']
+    boot_tags = ['kernel', 'initrd', 'cmdline', 'loader', 'nvram']
```
Comment (cbosdo) [inline salt/modules/virt.py:1935]: You need to be careful here since in the updated XML you may not have the nvram as this is computed

**Review by cbosdo — CHANGES_REQUESTED** [review]
Thanks for the work. This is a good first attempt, but I left a few comments to be addressed. You will also need to add unit test cases for that

File: salt/modules/virt.py [hunk 6]
```diff
@@ -1496,6 +1496,7 @@ def init(name,
     .. _disk element: https://libvirt.org/formatdomain.html#elementsDisks
     .. _graphics element: https://libvirt.org/formatdomain.html#elementsGraphics
     '''
+
```
Comment (cbosdo) [inline salt/modules/virt.py:1500]:
```suggestion
```

File: salt/modules/virt.py (hunk 4 above)
Comment (MeAndTheFirefly) [reply salt/modules/virt.py:1395]: I didn't know when did this line get inserted. I will go git diff and check carefully every time before I commit

Comment (cbosdo) [reply salt/modules/virt.py:1395]: happens :) but you'll need to add some doc for the new boot properties that you are adding

**Pull request closed without merging**